### PR TITLE
add authentication user test cases

### PR DESCRIPTION
**Pull request by HarithBashar**
I have add the test cases and edit the use case by adding some empty preventing statements.

### CODE REVIEW

File: src/main/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCase.kt [hunk 1]
```diff
@@ -7,6 +7,15 @@ class AuthenticateUserLoginUseCase(
...
     operator fun invoke(username: String, password: String): Result<User> {
+        if (username.isBlank()) {
+            return Result.failure(IllegalArgumentException("Username cannot be empty"))
+        }
+        if (password.isBlank()) {
+            return Result.failure(IllegalArgumentException("Password cannot be empty"))
+        }
+        if (!username.matches(Regex("^[A-Za-z0-9+_.-]+@(.+)$"))) {
+            return Result.failure(IllegalArgumentException("Invalid email format"))
+        }
```
Comment (yusufnasserdev) [inline src/main/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCase.kt]: We don't have an email as the username in the system.

**Comment by yusufnasserdev** [comment]
Please follow squad rules for naming branches next time.

File: src/main/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCase.kt [hunk 2]
```diff
@@ -7,6 +7,15 @@ class AuthenticateUserLoginUseCase(
     private val authRepo: AuthenticationRepository
 ) {
     operator fun invoke(username: String, password: String): Result<User> {
+        if (username.isBlank()) {
+            return Result.failure(IllegalArgumentException("Username cannot be empty"))
+        }
+        if (password.isBlank()) {
+            return Result.failure(IllegalArgumentException("Password cannot be empty"))
+        }
```
Comment (yusufnasserdev) [inline src/main/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCase.kt]: This should be in the UI instead, as it's input validation, not username and password validation.

Kindly refactor `LoginView` to include this instead.

File: src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import kotlin.test.assertEquals
+
+class AuthenticateUserLoginUseCaseTest {
+
+    private lateinit var authRepo: AuthenticationRepository
+    private lateinit var authenticateUserLogin: AuthenticateUserLoginUseCase
+
+    @BeforeEach
+    fun setUp() {
+        authRepo = mockk()
```
Comment (yusufnasserdev) [inline src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt]: Use `mockk(relaxed = true)` to avoid unpredictable tests behavior.

File: src/main/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCase.kt (hunk 1 above)
Comment (yusufnasserdev) [inline src/main/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCase.kt]: This is only a login function, we don't care what the input is as long as it's valid (decided by the UI), we simply query the given input against our database.

This kind of validation can be benefited from in a register-like use-case.

File: src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt
Comment (yusufnasserdev) [inline src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt:1]: When trying to assert an exception, simply use `assertThrows`, most of these usecases try to test something the use-case shouldn't be doing, please do the requested refactor first then re-request a review.

**Review by yusufnasserdev — CHANGES_REQUESTED** [review]

File: src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        val email = "test@example.com"
+        val password = "wrongpassword"
+        val exception = RuntimeException("Invalid credentials")
+        every { authRepo.login(email, password) } throws exception
+
+        // When
+        val result = authenticateUserLogin(email, password)
+
+        // Then
+        assertEquals(exception, result.exceptionOrNull())
```
Comment (Asmaa7071) [inline src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt]: you can use assertThrows directly

File: src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    @BeforeEach
+    fun setUp() {
+        authRepo = mockk()
+        authenticateUserLogin = AuthenticateUserLoginUseCase(authRepo)
+    }
+
+    @Test
+    fun `login should return success result when credentials are valid`() {
+        // Given
+        val email = "test@example.com"
```
Comment (Asmaa7071) [inline src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt]: login with user name not email

File: src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    fun `login should return failure result when both username and password are empty`() {
+        // Given
+        val username = ""
+        val password = ""
+        val exception = RuntimeException("Invalid credentials")
+        every { authRepo.login(username, password) } throws exception
+
+        // When & Then
+        assertThrows<RuntimeException> {
+            authenticateUserLogin(username, password).getOrThrow()
```
Comment (Asmaa7071) [inline src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt]: why you need to use .getOrThrow()

File: src/main/kotlin/ui/views/LoginView.kt [hunk 7]
```diff
@@ -11,7 +11,16 @@ class LoginView(
     fun getLoggedUser(): User {
         while (true) {
             val usernameInput = inputReader.readString("Username:")
+            if (usernameInput.isBlank()) {
```
Comment (Asmaa7071) [inline src/main/kotlin/ui/views/LoginView.kt]: you didn't need to check useName is blank cause  inputReader.readString already doing this

File: src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    private lateinit var authenticateUserLogin: AuthenticateUserLoginUseCase
+
+    @BeforeEach
+    fun setUp() {
+        authRepo = mockk(relaxed = true)
+        authenticateUserLogin = AuthenticateUserLoginUseCase(authRepo)
+    }
+
+    @Test
+    fun `login should return success result when credentials are valid`() {
```
Comment (Asmaa7071) [inline src/test/kotlin/logic/usecase/auth/AuthenticateUserLoginUseCaseTest.kt]: you can use Parameterized test here

**Review by yusufnasserdev — APPROVED** [review]
